src/sentry/static/sentry/app/components/substances/uploadSamples.jsx

```diff
@@ -0,0 +1,192 @@
+import React from 'react';
+import createReactClass from 'create-react-class';
+import Modal from 'react-bootstrap/lib/Modal';
+
+import {t} from 'app/locale';
+import ApiMixin from 'app/mixins/apiMixin';
+import IndicatorStore from 'app/stores/indicatorStore';
+import {FormState} from 'app/components/forms';
+
+const UploadSamplesButton = createReactClass({
+  displayName: 'UploadSamplesButton',
+
+  propTypes: {
+    disabled: PropTypes.bool,
+    style: PropTypes.object,
+    tooltip: PropTypes.string,
+    buttonTitle: PropTypes.string,
+    query: PropTypes.string,
+  },
+
+  mixins: [ApiMixin],
+
+  getInitialState() {
+    return {
+      isModalOpen: false,
+      formData: {
+        query: this.props.query,
+      },
+      errors: {},
+      value: null,
+      workflowVars: null,
+      workflowVars2: null,
+      setProcessVariables: null,
+      selectedFile: null,
+      loaded: 0,
+    };
+  },
+
+  onVariableChange(data) {
+    this.setState({
+      setProcessVariables: data,
+    });
+  },
+
+  onToggle() {
+    if (this.props.disabled) {
+      return;
+    }
+    this.setState({
+      isModalOpen: !this.state.isModalOpen,
+      state: FormState.READY,
+      formData: {
+        query: this.props.query,
+      },
+    });
+  },
+
+  onFieldChange(name, value) {
+    const formData = this.state.formData;
+    formData[name] = value;
+    this.setState({
+      formData,
+    });
+  },
+
+  onDefaultChange(e) {
+    this.onFieldChange('isDefault', e.target.checked);
+  },
+
+  onUserDefaultChange(e) {
+    this.onFieldChange('isUserDefault', e.target.checked);
+  },
+
+  handleSubmitSucces(user) {
+    //updateUser(user);
+    //this.setState({user});
+  },
+
+  handleSelectedFile(event) {
+    this.setState({
+      selectedFile: event.target.files[0],
+      loaded: 0,
+    });
+  },
+
+  handleUpload(event) {
+    if (!this.state.selectedFile) {
+      return;
+    }
+
+    this.setState({state: FormState.SAVING}, () => {
+      const loadingIndicator = IndicatorStore.add(t('Saving changes..'));
+
+      const endpoint = '/organizations/lab/substances/files/';
+
+      const reader = new FileReader();
+      reader.readAsBinaryString(this.state.selectedFile);
+
+      reader.onload = function() {
+        const data = {
+          content: btoa(reader.result),
+          filename: this.state.selectedFile.name,
+        };
+
+        this.api.request(endpoint, {
+          method: 'POST',
+          data,
+          success: response => {
+            this.onToggle();
+            this.setState({
+              state: FormState.READY,
+              errors: {},
+            });
+          },
+          error: err => {
+            let errors = err.responseJSON || true;
+            errors = errors.detail || true;
+            this.setState({
+              state: FormState.ERROR,
+              errors,
+            });
+          },
+          complete: () => {
+            IndicatorStore.remove(loadingIndicator);
```

Comment:
The loading indicator does not disappear after the samples have been uploaded, so I suspect that something is not working as expected here.

Reply:
Right, I'm not sure I changed anything here, but could be some interaction between components. It can be fixed later.